### ISSUE TITLE
Difference in detokenizer when the input “image_pil“ and “image_embeds“

### ISSUE THREAD

**Issue by qxhhou** [AILab-CVC/SEED-X#14]
In adapter_modules.py 
`    
    @torch.inference_mode()
    def get_image_embeds(self, image_pil=None, image_tensor=None, image_embeds=None, return_negative=True, image_size=448):
        assert int(image_pil is not None) + int(image_tensor is not None) + int(image_embeds is not None) == 1

        if image_pil is not None:
            image_tensor = self.image_transform(image_pil).unsqueeze(0).to(self.device, dtype=self.dtype)

        if image_tensor is not None:
            if return_negative:
                image_tensor_neg = torch.zeros_like(image_tensor)
                image_tensor = torch.cat([image_tensor, image_tensor_neg], dim=0)

            image_embeds = self.visual_encoder(image_tensor)
        elif return_negative:
            image_tensor_neg = torch.zeros(1, 3, image_size, image_size).to(image_embeds.device, dtype=image_embeds.dtype)
            image_embeds_neg = self.visual_encoder(image_tensor_neg)
            if self.vit_down:
                image_embeds_neg = image_embeds_neg.permute(0, 2, 1) # NLD -> NDL
                image_embeds_neg = F.avg_pool1d(image_embeds_neg, kernel_size=4, stride=4)
                image_embeds_neg = image_embeds_neg.permute(0, 2, 1)
            image_embeds = torch.cat([image_embeds, image_embeds_neg], dim=0)

        if self.discrete_model is not None:
            image_embeds = self.discrete_model.encode_image_embeds(image_embeds)
        image_embeds, pooled_image_embeds = self.encode_image_embeds(image_embeds)

        if return_negative:
            image_embeds, image_embeds_neg = image_embeds.chunk(2)
            pooled_image_embeds, pooled_image_embeds_neg = pooled_image_embeds.chunk(2)

        else:
            image_embeds_neg = None
            pooled_image_embeds_neg = None

        return image_embeds, image_embeds_neg, pooled_image_embeds, pooled_image_embeds_neg

 

`

It seems that when the input is '**image_pil**' It does not use the "**vit_down**"  the image_embeds shape is **(256, 4096)**

But  **MLLM output** image_embeds shape is **(64,4096)**

Althought after the Resampler layer, the output shape is always (64,*) .   there may be a mismatch？？

**Comment by geyuying** [comment]
Hi, during inference, we always use 'image_tensor', so 'image_pil' is None. When you use 'image_pil', you need to revise the code to use 'vit down'.
